specifyweb/frontend/js_src/lib/components/QueryComboBox/index.tsx

```diff
@@ -171,30 +172,42 @@ export function QueryComboBox({
            */
           field.isDependent())
           ? resource
-              .rgetPromise<string, AnySchema>(field.name)
-              .then(async (resource) =>
-                resource === undefined || resource === null
-                  ? {
-                      label: localized(''),
-                      resource: undefined,
-                    }
-                  : (value === formattedRef.current?.value &&
+              .rgetPromise<string, AnySchema>(field.name, true, false)
+              .then(async (resource) => {
+                setValidation([]);
+                if (resource === undefined || resource === null) {
+                  return {
+                    label: localized(''),
+                    resource: undefined,
+                  };
+                } else {
+                  const formatted =
+                    value === formattedRef.current?.value &&
                     typeof formattedRef.current === 'object'
-                      ? Promise.resolve(formattedRef.current.formatted)
-                      : format(
+                      ? await Promise.resolve(formattedRef.current.formatted)
+                      : await format(
                           resource,
                           typeof typeSearch === 'object'
                             ? typeSearch.formatter
                             : undefined,
                           true
-                        )
-                    ).then((formatted) => ({
-                      label:
-                        formatted ??
-                        naiveFormatter(field.relatedTable.label, resource.id),
-                      resource,
-                    }))
-              )
+                        );
+
+                  return {
+                    label:
+                      formatted ??
+                      naiveFormatter(field.relatedTable.label, resource.id),
+                    resource,
+                  };
+                }
+              })
+              .catch((_) => {
+                setValidation([formsText.invalidValue()]);
+                return {
+                  label: localized(''),
+                  resource: undefined,
+                };
+              })
```

Comment:
As of now, the code never reaches this catch block as the 404 response is not considered to be an error. There is no 404 dialog in RSS Export Feed anymore (context: #4955) and the QCBX instead displays the formatted invalid resource and pretends it as valid. 

We can hit the catch block if we also throw an error when `status === Http.NOT_FOUND` in line 66 over here:
https://github.com/specify/specify7/blob/6a364e00f96d5c81f0853c65b3d12ed0918fc202/specifyweb/frontend/js_src/lib/utils/ajax/backboneAjax.ts#L64-L69
In which case, the error gets caught in QCBX, value gets cleared and a save blocker is set but somehow it does not trigger an unload protect and the blocker is not visible until you click on the field. My guess is there needs to be a re-render in the parent component for the unload protect to trigger.

Reply:
after adding hijackBackboneAjax like described in https://github.com/specify/specify7/pull/5139#discussion_r1690614195, it looks like the save blocker is set correctly and the save buttons turn red:

![Screenshot 2024-07-24 at 17 17 51](https://github.com/user-attachments/assets/dc189a69-4910-4b5e-a7d4-7c18d354bf67)

could you verify if you also get this behavior after making that change?

Reply:
Hmm, save does get blocked correctly in the CO form. However, in a prod environment the page redirects to a 404 page, which doesn't solve the issue in #4955.

Save blocker still doesn't work correctly in RSS export feed although the values did get cleared
![image](https://github.com/user-attachments/assets/808296b2-6bed-46c6-946b-bcc6dbf96a34)
and the blocker is only visible after clicking on the field --- but it doesn't matter as in prod the page will still redirect to 404
![image](https://github.com/user-attachments/assets/44a4e1bc-dcae-4df3-9a96-f607c3e629b5)

Making the change here: https://github.com/specify/specify7/pull/5139#discussion_r1690241121 along with wrapping `hijackBackboneAjax` over the resource fetch directly in `resourceApi.ts` helps avoid the 404

Reply:
I believe if we use `hijackBackboneAjax` higher up the call stack as opposed to wrapping it over `resource.fetch()`, it gets set as undefined before all chained promises are resolved
https://github.com/specify/specify7/blob/8d0052c7a8f37a1f2eb14a1e730279bd117a7608/specifyweb/frontend/js_src/lib/utils/ajax/backboneAjax.ts#L29

and then the `expectedErrors` over here is an empty list instead of `[404]` and so it gets treated as an uncaught error
https://github.com/specify/specify7/blob/8d0052c7a8f37a1f2eb14a1e730279bd117a7608/specifyweb/frontend/js_src/lib/utils/ajax/response.ts#L11-L26

Reply:
> Hmm, save does get blocked correctly in the CO form. However, in a prod environment the page redirects to a 404 page, which doesn't solve the issue in https://github.com/specify/specify7/issues/4955.

ah, sorry! I forgot that the behavior is different in production

---

You are correct
The cause of the issue is that rget is actual awaiting two promises:
1. Loading the resource itself (even if it's already loaded, this still creates a promise):
2. Loading the related resource:
https://github.com/specify/specify7/blob/9faedc70ac8cb78dbd676a840770e8ebf1476500/specifyweb/frontend/js_src/lib/components/DataModel/resourceApi.ts#L521-L522

the assumption that `hijackBackboneAjax()` makes is that the variables it sets (`expectedErrors`, among others) are going to be retrieved in the same cycle - but that is not the case as the rget has two promises, and the one we are interested in comes 2nd.

The issue is that `hijackBackboneAjax()` is called concurrently by multiple query combo boxes (i.e accession), and so if accession resource is fetched earlier, the `hijackBackboneAjax()` does cleanup afterward, unsettling `expectedErrors` and other variables - before the collector request was sent out with correct `expectedErrors`

In all other places where hijackBackboneAjax is used, it's used directly on `.fetch()`, so this issue is not present.

two options:
- 1st (not great): make rget only call fetch() if resource is not yet fetched. not great as then this bug would still be present when working with resource that is not yet fetched
- 2nd (close to what you have right now): hijackBackboneAjax itself is a hack. relying on global variables is not great, and even worse when working with async actions. a better solution is to pass fetch options along with requests that do the fetching.

we don't want to do too many changes to resourceApi.ts as we are hoping to do major refactor to it soon, but in an ideal world the change you made with introducing the "options" parameter would be made for all methods in resourceApi that make network requests.

for now, could you please make the following changes:
- only call hijackBackboneAjax() in .fetch() if some options were actually provided - a performance optimization since .fetch() is very much on a hot path
- post a comment in https://github.com/specify/specify7/issues/4286 summarising the issue ("we need more control over how resourceApi makes network requests ...") and say that the new resourceApi API we will design should have fetching options on all async methods. since the new API will be using `ajax()` directly, it will be able to easily pass on the options, without the need for hacky `hijackBackboneAjax()`